src/expr.cpp

```diff
@@ -3183,18 +3199,23 @@ llvm::Value *AssignExpr::GetValue(FunctionEmitContext *ctx) const {
         return value;
     }
     case MulAssign:
+    case AddAssign:
+    case SubAssign: {
+        // This should be caught during type checking
+        AssertPos(pos, !CastType<ArrayType>(type) && !CastType<StructType>(type));
+        const auto wrapSemantics = lvalue->GetType()->IsSignedType() ? WrapSemantics::NSW : WrapSemantics::None;
```

Comment:
https://godbolt.org/z/4GqvxT4PW

In case of `*=`, `+=`, and `-=` the decision about type of operation should be done in the same way as it's done for `*`, `+`, and `-`, i.e. the common type of operation should be computed with the same rules as for binary operation alone. `signed` + `unsigned` yields `unsigned` operation, but just based on the `lvalue` type.

Reply:
Updated to check both l and r values for signedness.